src/mlpack/core/tree/rectangle_tree/minimal_coverage_sweep_impl.hpp

```diff
@@ -0,0 +1,260 @@
+    {
+      if (lowerBound1[k] > highBound2[k] || lowerBound2[k] > highBound2[k])
+        overlappedArea *= 0;
+      else
+        overlappedArea *= std::min(highBound1[k], highBound2[k]) -
+            std::max(lowerBound1[k], lowerBound2[k]);
+    }
+  }
+
+  return area1 + area2 - overlappedArea;
+}
+
+template<typename SplitPolicy>
+template<typename TreeType>
+typename TreeType::ElemType MinimalCoverageSweep<SplitPolicy>::
+SweepLeafNode(const size_t axis,
+              const TreeType* node,
+              typename TreeType::ElemType& axisCut)
+{
+  typedef typename TreeType::ElemType ElemType;
+
+  std::vector<SortStruct<ElemType>> sorted(node->Count());
+
+  sorted.resize(node->Count());
+
+  for (size_t i = 0; i < node->NumPoints(); i++)
+  {
+    sorted[i].d = node->Dataset().col(node->Point(i))[axis];
+    sorted[i].n = i;
+  }
+
+  // Sort high bounds of children.
+  std::sort(sorted.begin(), sorted.end(), StructComp<ElemType>);
+
+  size_t splitPointer = fillFactor * node->Count();
+
+  axisCut = sorted[splitPointer - 1].d;
+
+  // Check if the partition is suitable.
+  if (!CheckLeafSweep(node, axis, axisCut))
+    return std::numeric_limits<ElemType>::max();
+
+  std::vector<ElemType> lowerBound1(node->Bound().Dim());
+  std::vector<ElemType> highBound1(node->Bound().Dim());
+  std::vector<ElemType> lowerBound2(node->Bound().Dim());
+  std::vector<ElemType> highBound2(node->Bound().Dim());
+
+  // Find lower and high bounds of two resulting nodes.
+  for (size_t k = 0; k < node->Bound().Dim(); k++)
+  {
+    lowerBound1[k] = node->Dataset().col(node->Point(sorted[0].n))[k];
+    highBound1[k] = node->Dataset().col(node->Point(sorted[0].n))[k];
+
+    for (size_t i = 1; i < splitPointer; i++)
+    {
+      if (node->Dataset().col(node->Point(sorted[i].n))[k] < lowerBound1[k])
+        lowerBound1[k] = node->Dataset().col(node->Point(sorted[i].n))[k];
+      if (node->Dataset().col(node->Point(sorted[i].n))[k] > highBound1[k])
+        highBound1[k] = node->Dataset().col(node->Point(sorted[i].n))[k];
+    }
+
+    lowerBound2[k] = node->Dataset().col(
+        node->Point(sorted[splitPointer].n))[k];
+    highBound2[k] = node->Dataset().col(node->Point(sorted[splitPointer].n))[k];
+
+    for (size_t i = splitPointer + 1; i < node->NumChildren(); i++)
+    {
+      if (node->Dataset().col(node->Point(sorted[i].n))[k] < lowerBound2[k])
+        lowerBound2[k] = node->Dataset().col(node->Point(sorted[i].n))[k];
+      if (node->Dataset().col(node->Point(sorted[i].n))[k] > highBound2[k])
+        highBound2[k] = node->Dataset().col(node->Point(sorted[i].n))[k];
+    }
+  }
+
+  // Evaluate the cost of the split i.e. calculate the total coverage
+  // of two resulting nodes.
+
+  ElemType area1 = 1.0, area2 = 1.0;
+  ElemType overlappedArea = 1.0;
+
+  for (size_t k = 0; k < node->Bound().Dim(); k++)
+  {
+    area1 *= highBound1[k] - lowerBound1[k];
+    area2 *= highBound2[k] - lowerBound2[k];
+  }
```

Comment:
I think this function could be simplified a lot if you used the native `HRectBound` functions.  Like instead of creating `lowerBound1`, `lowerBound2`, `highBound1`, and `highBound2` as `std::vector<ElemType>`s, you could just create `bound1` and `bound2` as `HRectBound<MetricType, ElemType>` and then use `operator|=()` to expand the bounds as necessary, then `Volume()` to calculate the last bit.

Also it seems like `overlappedArea` is not used here or calculated, so maybe that is wrong, or maybe it just needs to be removed.  I thought that in the R+ and R++ tree, nodes could not overlap, so I am not sure why `overlappedArea` is not 0 for both leaf nodes and non-leaf nodes.